components/range_cache_memory_engine/src/range_stats.rs

```diff
@@ -381,15 +337,139 @@ impl RangeStatsManager {
+        // amplification
+        let avg_mvcc_amplification = regions_activity
+            .iter()
+            .map(|(_, ra)| ra.cop_detail.mvcc_amplification())
+            .sum::<f64>()
+            / regions_activity.len() as f64;
+
+        let evict_candidates: Vec<_> = {
+            let mut regions_loaded = self.region_loaded_at.write().unwrap();
+            let mut evict_candidates = regions_activity;
+            info!(
+                "IME evict candidate count before filter";
+                "count" => evict_candidates.len(),
+            );
+            let mut filter_by_time = 0;
+            let mut filter_by_mvcc_amplification = 0;
+            evict_candidates.retain(|(r, ra)| {
+                let time_loaded = regions_loaded.entry(r.id).or_insert(Instant::now());
+                // Do not evict regions with high mvcc amplification
+                if ra.cop_detail.mvcc_amplification() > avg_mvcc_amplification {
+                    filter_by_mvcc_amplification += 1;
+                    return false;
+                }
+                let now = Instant::now();
+                // Do not evict regions that are loaded recently
+                if now - *time_loaded < self.evict_min_duration {
+                    filter_by_time += 1;
+                    return false;
+                }
+                true
+            });
+            evict_candidates.sort_by(|a, b| {
+                let next_prev_a = a.1.cop_detail.iterated_count();
+                let next_prev_b = b.1.cop_detail.iterated_count();
+                next_prev_a.cmp(&next_prev_b)
+            });
+
+            info!(
+                "IME evict candidate count after filter";
+                "count" => evict_candidates.len(),
+                "filter_by_time" => filter_by_time,
+                "filter_by_mvcc_amplification" => filter_by_mvcc_amplification,
+            );
+
+            evict_candidates.into_iter().map(|(r, _)| r).collect()
+        };
+        // Evict two regions each time to reduce the probability that an un-dropped
+        // ongoing snapshot blocks the process
+        for regions in evict_candidates.chunks(2) {
+            let mut deleteable_regions = vec![];
+
+            let (tx, mut rx) = mpsc::channel(3);
+            for r in regions {
+                info!(
+                    "ime evict on soft limit reached";
+                    "region_to_evict" => ?r,
+                );
+
+                let tx_clone = tx.clone();
+                deleteable_regions.extend(evict_region(
+                    &CacheRegion::from_region(r),
+                    EvictReason::MemoryLimitReached,
+                    // This callback will be executed when eviction finishes at `on_delete_regions`
+                    // and when the reletive rx.recv() returns, we know some memory are freed.
+                    Some(Box::new(move || {
+                        Box::pin(async move {
```

Comment:
Or I can block_on here, so that fn is till FnOnce() -> () rather than FnOnce() -> Pin<Box<dyn Future<Output = ()> + Send>>. What do you think? @glorv

Reply:
Ok, you can keep the current logic, I don't notice you changed to future channel.
I  original thought to just add a sync wait with timeout is enough to avoiding blocking forever.